src/hotspot/cpu/aarch64/aarch64_neon.ad

```diff
@@ -5683,39 +5683,91 @@ instruct vround2D_reg(vecX dst, vecX src, immI rmode) %{
   ins_pipe(vdop_fp128);
 %}
 
-instruct vpopcount4I(vecX dst, vecX src) %{
-  predicate(UsePopCountInstruction && n->as_Vector()->length() == 4);
+instruct vpopcountID(vecD dst, vecD src) %{
+  predicate(n->as_Vector()->length_in_bytes() < 16);
   match(Set dst (PopCountVI src));
-  format %{
-    "cnt     $dst, $src\t# vector (16B)\n\t"
-    "uaddlp  $dst, $dst\t# vector (16B)\n\t"
-    "uaddlp  $dst, $dst\t# vector (8H)"
+  ins_cost(3 * INSN_COST);
+  format %{ "vpopcountI  $dst, $src\t# vector (8B/4H/2S)" %}
+  ins_encode %{
+    assert(UsePopCountInstruction, "unsupported");
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    __ cnt(as_FloatRegister($dst$$reg), __ T8B,
+           as_FloatRegister($src$$reg));
+    if (bt == T_SHORT || bt == T_INT) {
+      __ uaddlp(as_FloatRegister($dst$$reg), __ T8B,
+                as_FloatRegister($dst$$reg));
+    }
+    if (bt == T_INT) {
```

Comment:
I find it hard to reason about the code in its current form.

Maybe make the second `if` (`bt == T_INT`) nested and move it under  `if (bt == T_SHORT || bt == T_INT)`?